When inviting new member, can set initial role + teams

benvinegar opened the pull request:
Before you could only *adjust* a user's role / team membership **after** they had already been invited – which often resulted in new members being added with no initial teams being assigned. 

This can lead to a confusing experience when first using Sentry, since no teams means no projects means no errors to browse.

/cc @getsentry/team @dcramer

Before:

![image](https://cloud.githubusercontent.com/assets/2153/18296431/7b084d1a-745e-11e6-915d-2ecab2e98697.png)

After:

![image](https://cloud.githubusercontent.com/assets/2153/18296417/685c75d8-745e-11e6-888b-c7c49e980dfd.png)

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/getsentry/sentry/4089)
<!-- Reviewable:end -->

benvinegar commented on src/sentry/web/frontend/create_organization_member.py:
```diff
@@ -16,20 +17,34 @@
         if settings.SENTRY_ENABLE_INVITES:
             form_cls = InviteOrganizationMemberForm
         else:
             form_cls = AddOrganizationMemberForm
```
I need to verify this works with `AddOrganizationMemberForm` (probably not) – this has only been tested with `InviteOrganizationMemberForm`.

benvinegar commented:
cc @mitsuhiko @matt – would be grateful to get eyes on this

mattrobenolt commented:
Reviewed 4 of 8 files at r1, 8 of 9 files at r2.
Review status: 12 of 13 files reviewed at latest revision, 5 unresolved discussions.

---

*[src/sentry/web/forms/base_organization_member.py, line 12 \[r2\]](https://reviewable.io:443/reviews/getsentry/sentry/4089#-KRA4rRzFdYEMOo7i0Si:-KRA4rS07p9PQB3TveEO:bygd3zc) ([raw file](https://github.com/getsentry/sentry/blob/e36461ee4a4afe496dbf2f7076b9200b2f2012be/src/sentry/web/forms/base_organization_member.py#L12)):*
> ```Python
> 
> 
> class BaseOrganizationMember(forms.ModelForm):
> ```

nitpicking, but I'd name this `*Form` so it's clear by convention.

After reading more, this actually was confusing above. Since it wasn't clear that `AddOrganizationMemberForm` subclassed a form.

---

*[src/sentry/web/forms/base_organization_member.py, line 41 \[r2\]](https://reviewable.io:443/reviews/getsentry/sentry/4089#-KRA5wOROm3wxzxtS_mH:-KRA5wOROm3wxzxtS_mI:b-6l5ndn) ([raw file](https://github.com/getsentry/sentry/blob/e36461ee4a4afe496dbf2f7076b9200b2f2012be/src/sentry/web/forms/base_organization_member.py#L41)):*
> ```Python
>         self.fields['teams'].queryset = all_teams
> 
>     def save_team_assignments(self, organization_member):
> ```

If we're intending to just 100% replace all existing memberships, we can do this more efficiently:

```python
@transaction.atomic
def save_team_assignments(self, organization_member):
    OrganizationMemberTeam.objects.filter(organizationmember=organization_member).delete()
    OrganizationMemberTeam.objects.bulk_create([
        OrganizationMemberTeam(team=team, organizationmember=organization_member)
        for team in self.cleaned_data['teams']
    ])
```

This will just delete and recreate all new in 2 queries within a single transaction, rather than O(n) queries.

Now moving along, I see this was done before and just copy/pasted. So I guess it's ok to stay, but this is an improvement regardless. I won't block on it though since it pre-existed.

---

*[src/sentry/web/frontend/create_organization_member.py, line 42 \[r2\]](https://reviewable.io:443/reviews/getsentry/sentry/4089#-KRA9DfkKNUIDfji7nyY:-KRA9DfkKNUIDfji7nyZ:b-ahm88j) ([raw file](https://github.com/getsentry/sentry/blob/e36461ee4a4afe496dbf2f7076b9200b2f2012be/src/sentry/web/frontend/create_organization_member.py#L42)):*
> ```Python
> 
>         all_teams = Team.objects.filter(
>             organization=organization,
> ```

This should be scoped to just `VISIBLE` teams. Just add `status=TeamStatus.VISIBLE` to the queryset.

---

*[src/sentry/web/frontend/create_organization_member.py, line 45 \[r2\]](https://reviewable.io:443/reviews/getsentry/sentry/4089#-KRA98Gxy1T-H5Z_mbOr:-KRA98Gxy1T-H5Z_mbOs:brk7f5n) ([raw file](https://github.com/getsentry/sentry/blob/e36461ee4a4afe496dbf2f7076b9200b2f2012be/src/sentry/web/frontend/create_organization_member.py#L45)):*
> ```Python
>         )
> 
>         # TODO if not can admim
> ```

This seems like an import TODO.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/getsentry/sentry/4089)*
<!-- Sent from Reviewable.io -->

benvinegar commented:
@matt – thanks for the feedback. I think I've addressed all the required changes if you don't mind looking again.

---

Reviewed 1 of 9 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[src/sentry/web/forms/base_organization_member.py, line 12 \[r2\]](https://reviewable.io:443/reviews/getsentry/sentry/4089#-KRA4rRzFdYEMOo7i0Si:-KRAGsnhkwUb6-g-2Xu8:bt1b19p) ([raw file](https://github.com/getsentry/sentry/blob/e36461ee4a4afe496dbf2f7076b9200b2f2012be/src/sentry/web/forms/base_organization_member.py#L12)):*
<details><summary><i>Previously, mattrobenolt (Matt Robenolt) wrote…</i></summary>
> nitpicking, but I'd name this `*Form` so it's clear by convention.
> 
> After reading more, this actually was confusing above. Since it wasn't clear that `AddOrganizationMemberForm` subclassed a form.
</details>
Renamed.

---

*[src/sentry/web/forms/base_organization_member.py, line 41 \[r2\]](https://reviewable.io:443/reviews/getsentry/sentry/4089#-KRA5wOROm3wxzxtS_mH:-KRAGujJdH1wNLXrcDMx:bm68jeq) ([raw file](https://github.com/getsentry/sentry/blob/e36461ee4a4afe496dbf2f7076b9200b2f2012be/src/sentry/web/forms/base_organization_member.py#L41)):*
<details><summary><i>Previously, mattrobenolt (Matt Robenolt) wrote…</i></summary>
> If we're intending to just 100% replace all existing memberships, we can do this more efficiently:
> 
> ```python
> @transaction.atomic
> def save_team_assignments(self, organization_member):
>     OrganizationMemberTeam.objects.filter(organizationmember=organization_member).delete()
>     OrganizationMemberTeam.objects.bulk_create([
>         OrganizationMemberTeam(team=team, organizationmember=organization_member)
>         for team in self.cleaned_data['teams']
>     ])
> ```
> 
> This will just delete and recreate all new in 2 queries within a single transaction, rather than O(n) queries.
> 
> Now moving along, I see this was done before and just copy/pasted. So I guess it's ok to stay, but this is an improvement regardless. I won't block on it though since it pre-existed.
</details>
I'm going to leave this as-is for now. I'll do the follow-up though.

---

*[src/sentry/web/frontend/create_organization_member.py, line 29 \[r1\]](https://reviewable.io:443/reviews/getsentry/sentry/4089#-KR1DQ9pxCHzqZutaP1J-r1-29:-KRAH0nvh7PXMIC8FQjv:bt2vr1b) ([raw file](https://github.com/getsentry/sentry/blob/318bc0b5c784bf8aa956815de3c4ba90311432e1/src/sentry/web/frontend/create_organization_member.py#L29)):*
<details><summary><i>Previously, benvinegar (Ben Vinegar) wrote…</i></summary>
> I need to verify this works with `AddOrganizationMemberForm` (probably doesn't) – this has only been tested with `InviteOrganizationMemberForm`.
</details>
Fixed.

---

*[src/sentry/web/frontend/create_organization_member.py, line 42 \[r2\]](https://reviewable.io:443/reviews/getsentry/sentry/4089#-KRA9DfkKNUIDfji7nyY:-KRAH3qKnMBa11QCC5xe:b-896fix) ([raw file](https://github.com/getsentry/sentry/blob/e36461ee4a4afe496dbf2f7076b9200b2f2012be/src/sentry/web/frontend/create_organization_member.py#L42)):*
<details><summary><i>Previously, mattrobenolt (Matt Robenolt) wrote…</i></summary>
> This should be scoped to just `VISIBLE` teams. Just add `status=TeamStatus.VISIBLE` to the queryset.
</details>
Done.

---

*[src/sentry/web/frontend/create_organization_member.py, line 45 \[r2\]](https://reviewable.io:443/reviews/getsentry/sentry/4089#-KRA98Gxy1T-H5Z_mbOr:-KRAH53FHMaOctUNkWSu:bw1l0hf) ([raw file](https://github.com/getsentry/sentry/blob/e36461ee4a4afe496dbf2f7076b9200b2f2012be/src/sentry/web/frontend/create_organization_member.py#L45)):*
<details><summary><i>Previously, mattrobenolt (Matt Robenolt) wrote…</i></summary>
> This seems like an import TODO.
</details>
Removed. It's not appropriate here – it was borrowed from `organization_member_settings.py`, where, if you cannot admin a user, you instead see a "view" page rather than a form.

In the case of inviting a user, if you cannot admin the user, you are rejected from visiting this page entirely. I couldn't find a way to reach this page as a regular user.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/getsentry/sentry/4089#-:-KRAHWIAKcu_7s18Hqze:b-oit52o)*
<!-- Sent from Reviewable.io -->

mattrobenolt commented:
Reviewed 4 of 6 files at r3.
Review status: 11 of 13 files reviewed at latest revision, 2 unresolved discussions.

---

*[src/sentry/web/frontend/create_organization_member.py, line 39 \[r3\]](https://reviewable.io:443/reviews/getsentry/sentry/4089#-KRALTbA00V3NSb6dicK:-KRALTbBwjahayKMvz0W:b-8you8m) ([raw file](https://github.com/getsentry/sentry/blob/d11e9e735d9b94fe108361dd5db8772bc4640932/src/sentry/web/frontend/create_organization_member.py#L39)):*
> ```Python
> 
>     def handle(self, request, organization):
>         can_admin, allowed_roles = self.get_allowed_roles(request, organization)
> ```

If it's unused, can we do `_, allowed_rules = self.get_allowed_roles(request, organization)` to indicate it's unused explicitly.

Not entirely sure why this isn't a linter warning for an unused variable.

---

*[src/sentry/web/frontend/organization_member_settings.py, line 53 \[r3\]](https://reviewable.io:443/reviews/getsentry/sentry/4089#-KRALz8UB4AbZzCXWRd9:-KRALz8UB4AbZzCXWRdA:bp6vyhi) ([raw file](https://github.com/getsentry/sentry/blob/d11e9e735d9b94fe108361dd5db8772bc4640932/src/sentry/web/frontend/organization_member_settings.py#L53)):*
> ```Python
>         return self.redirect(redirect)
> 
>     def view_member(self, request, organization, member, all_teams):
> ```

You're passing in `all_teams`, but why aren't you using this value in the `context`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/getsentry/sentry/4089)*
<!-- Sent from Reviewable.io -->

benvinegar commented:
Review status: 11 of 13 files reviewed at latest revision, 2 unresolved discussions.

---

*[src/sentry/web/frontend/create_organization_member.py, line 39 \[r3\]](https://reviewable.io:443/reviews/getsentry/sentry/4089#-KRALTbA00V3NSb6dicK:-KRAMpIED4Wo4AksVQMF:b-7m2g9y) ([raw file](https://github.com/getsentry/sentry/blob/d11e9e735d9b94fe108361dd5db8772bc4640932/src/sentry/web/frontend/create_organization_member.py#L39)):*
<details><summary><i>Previously, mattrobenolt (Matt Robenolt) wrote…</i></summary>
> If it's unused, can we do `_, allowed_rules = self.get_allowed_roles(request, organization)` to indicate it's unused explicitly.
> 
> Not entirely sure why this isn't a linter warning for an unused variable.
</details>
`_` is declared in the upper scope already, so I left it as-is

---

*[src/sentry/web/frontend/organization_member_settings.py, line 53 \[r3\]](https://reviewable.io:443/reviews/getsentry/sentry/4089#-KRALz8UB4AbZzCXWRd9:-KRAMsE9NATwwPfsgfnL:b-p4ybu) ([raw file](https://github.com/getsentry/sentry/blob/d11e9e735d9b94fe108361dd5db8772bc4640932/src/sentry/web/frontend/organization_member_settings.py#L53)):*
<details><summary><i>Previously, mattrobenolt (Matt Robenolt) wrote…</i></summary>
> You're passing in `all_teams`, but why aren't you using this value in the `context`?
</details>
Oops, fixed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/getsentry/sentry/4089)*
<!-- Sent from Reviewable.io -->

mattrobenolt commented:
Reviewed 1 of 6 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/getsentry/sentry/4089)*
<!-- Sent from Reviewable.io -->
